src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,363 @@
+        public void AddToPendingBatch(ProvenBlockHeader provenBlockHeader, HashHeightPair newTip)
+        {
+            this.PendingBatch.AddOrUpdate(newTip.Height, provenBlockHeader, (key, value) => { return provenBlockHeader; });
+
+            this.Cache.AddOrUpdate(newTip.Height, provenBlockHeader, (key, value) => { return provenBlockHeader; });
+
+            this.pendingTipHashHeight = newTip;
+        }
+
+        /// <summary> Saves pending <see cref="ProvenBlockHeader"/> items to the <see cref="ProvenBlockHeaderRepository"/>.</summary>
+        private async Task SaveAsync()
+        {
+            if (this.PendingBatch.Count == 0)
+            {
+                this.logger.LogTrace("(-)[PENDING_BATCH_EMPTY]");
+                return;
+            }
+
+            if (!this.PendingBatch.Keys.OrderBy(s => s).SequenceEqual(this.PendingBatch.Keys))
+            {
+                this.logger.LogTrace("(-)[PENDING_BATCH_INCORRECT_SEQEUNCE]");
+                throw new InvalidOperationException("Invalid ProvenBlockHeader pending batch sequence - unable to save to the database repository.");
+            }
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddInsertTime(o)))
+            {
+                var pendingBatchKeys = this.PendingBatch.Keys.ToList();
+
+                List<ProvenBlockHeader> pendingBatchValues = this.PendingBatch.Values.ToList();
+
+                await this.provenBlockHeaderRepository.PutAsync(
+                    pendingBatchValues, this.pendingTipHashHeight).ConfigureAwait(false);
+
+                pendingBatchKeys.ForEach(
+                    blockHeight => this.PendingBatch.TryRemove(blockHeight, out ProvenBlockHeader unused));
+
+                this.pendingTipHashHeight = null;
+            }
+        }
+
+        /// <summary>
+        /// Gets <see cref="ProvenBlockHeader"></see> items from cache with specific key if present.
+        /// </summary>
+        /// <param name="key"> Item's key.</param>
+        /// <param name = "items"> Dictionary of items in cache</param>
+        /// <returns> <see cref="ProvenBlockHeader"></see> if cache contains the item, <c>null</c> otherwise.</returns>
+        private ProvenBlockHeader GetHeaderFromCache(int key, ConcurrentDictionary<int, ProvenBlockHeader> items)
+        {
+            if (items.TryGetValue(key, out ProvenBlockHeader header))
+            {
+                this.logger.LogTrace("(-):{0}", header);
+
+                return header;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Manages store cache size. Remove 30% of items if the memory cache size has been reached.
+        /// </summary>
+        private void ManangeCacheSize()
+        {
+            do
+            {
+                if (this.CacheSizeInBytes > this.MemoryCacheSizeLimitInBytes)
+                {
+                    List<int> sortedList = this.Cache.Keys.OrderBy(s => s).ToList();
```

Comment:
So you sort by height and delete items. It's not LRU cache then. 
Even without sort it won't be LRU because dictionaries don't support order.